File: 22从上到下打印二叉树/Solution.cpp
class Solution {
private:
    queue<TreeNode*>  point;
public:
    vector<int> PrintFromTopToBottom(TreeNode* root) {
        vector<int> res;
        point.push(root);
        while (!point.empty()){
            int len=point.size();
            while(len>0){
                len--;
                TreeNode* t=point.front();
                point.pop();
                if (t==NULL) continue;
                res.push_back(t->val);
                point.push(t->left);
                 point.push(t->right);
            }
        }
        return res;
    }
};
/* 用一个队列去解决这个问题是极为巧妙的，首先队列可以将所有的子节点按顺序
存入队列中，而且每打印一个队列中元素的时候总会将该元素的子节点按左右顺序
存入队列，这样保证了逐层从左向右的完整遍历 */

File: 22从上到下打印二叉树/Solution.java

public class Solution {
    private Queue<TreeNode> point=new LinkedList<TreeNode>();
    public ArrayList<Integer> PrintFromTopToBottom(TreeNode root) {
         ArrayList<Integer> res=new ArrayList<>();
        point.offer(root);
        while (!point.isEmpty()){
            int len=point.size();
            while(len>0){
                len--;
                TreeNode t=point.peek();
                point.poll();
                if (t==null) continue;
                res.add(t.val);
                point.offer(t.left);
                 point.offer(t.right);
            }
        }
        return res;
}
需要注意Queue和linkedlist 的关系和初始化问题，还有queue的offer 
poll  peek方法